src/lsm/scan_groove.zig

```diff
@@ -0,0 +1,797 @@
+
+pub fn ScanGrooveType(
+    comptime Groove: type,
+    comptime Storage: type,
+    comptime scan_max: comptime_int,
+) type {
+    comptime assert(std.meta.fields(Groove.IndexTrees).len > 0);
+
+    return struct {
+        const Self = @This();
+        const ScanBuffer = ScanBufferType(Storage);
+
+        pub const Scan = ScanType(Groove, Storage, scan_max);
+        pub const Fetcher = FetcherType(Scan, Groove, Storage);
+
+        slots: [scan_max]Scan = undefined,
+        slots_used: u32 = 0,
+
+        fetcher: Fetcher = undefined,
+
+        pub fn Tree(comptime field: std.meta.FieldEnum(Groove.IndexTrees)) type {
+            return std.meta.fieldInfo(Groove.IndexTrees, field).field_type;
+        }
+
+        pub fn CompositeKey(comptime field: std.meta.FieldEnum(Groove.IndexTrees)) type {
+            return Tree(field).Table.Key;
+        }
+
+        pub fn CompositeKeyPrefix(comptime field: std.meta.FieldEnum(Groove.IndexTrees)) type {
+            const Key = CompositeKey(field);
+            return std.meta.fieldInfo(Key, .field).field_type;
+        }
+
+        fn ScanImplType(comptime field: std.meta.FieldEnum(Scan.Dispatcher)) type {
+            return std.meta.fieldInfo(Scan.Dispatcher, field).field_type;
+        }
+
+        /// Initializes a Scan over the index specified by `index`, searching for an exact match.
+        /// Produces the criteria equivalent to
+        /// `WHERE field = $value`.
+        /// 
+        /// Results are sorted by `direction`.
+        pub fn equal(
+            self: *Self,
+            comptime index: std.meta.FieldEnum(Groove.IndexTrees),
+            buffer: *const ScanBuffer,
+            snapshot: u64,
+            value: CompositeKeyPrefix(index),
+            timestamp_range: TimestampRange,
+            direction: Direction,
+        ) *Scan {
+            const field = comptime index_to_dispatcher(index);
+            const ScanImpl = ScanImplType(field);
+            return scan_add(
+                self,
+                field,
+                ScanImpl.init(
+                    self.tree(index),
+                    buffer,
+                    snapshot,
+                    .{ .field = value, .timestamp = timestamp_range.min },
+                    .{ .field = value, .timestamp = timestamp_range.max },
+                    direction,
+                ),
+            );
+        }
+
+        /// Initializes a Scan over the index specified by `field`, searching for a range.
+        /// Produces the criteria equivalent to
+        /// `WHERE field BETWEEN $min AND $max` (inclusive range).
+        /// 
+        /// Results are not sorted.
+        pub fn between(
+            self: *Self,
+            comptime index: std.meta.FieldEnum(Groove.IndexTrees),
+            buffer: *const ScanBuffer,
+            snapshot: u64,
+            min: CompositeKeyPrefix(index),
+            max: CompositeKeyPrefix(index),
+            timestamp_range: TimestampRange,
```

Comment:
As a second thought, `TimestampRange` is a very bad abstraction because its main function, paging, is applicable only for `equal`  operations ... for `between` we should take the entire tuple `prefix, timestamp`.

Also, we don't ever need `{ .min, .max }` here. Such use-case should use `ScanTreeType` directly instead. For the level of abstraction `ScanGrooveType` provides for `equal`, only the "last timestamp seen" is enough, as we can infer if it's the lower or upper bound based on the `direction`.